exonum-java-binding/core/rust/src/testkit/mod.rs

```diff
@@ -216,88 +216,118 @@ fn create_java_keypair<'a>(
+        let configs = parse_service_specs(env, service_specs_obj, artifact_id, artifact_filename)?;
+        instance_configs.extend(configs);
+    }
+    Ok(instance_configs)
+}
 
-        let num_specs = env.get_array_length(service_specs_obj)?;
-
-        for k in 0..num_specs {
-            let service_spec = env.get_object_array_element(service_specs_obj, k)?;
-
-            let (service_id, service_name) = get_service_id_and_name(env, service_spec)?;
-            let config_params: jbyteArray = env
-                .get_field(service_spec, "configuration", "[B")?
-                .l()?
-                .into_inner();
-            let config = env.convert_byte_array(config_params)?;
+// Converts Java array of `ServiceSpec` instances into vector of `InstanceConfig` for specific artifact.
+fn parse_service_specs(
+    env: &JNIEnv,
+    specs_array: jobjectArray,
+    artifact_id: String,
+    artifact_filename: String,
+) -> JniResult<Vec<InstanceConfig>> {
+    let num_specs = env.get_array_length(specs_array)?;
 
-            let spec = InstanceSpec::new(service_id, service_name, &artifact_id)
-                .expect("Unable to create instance specification for service");
-            let cfg = InstanceConfig::new(spec, Some(artifact_filename.to_bytes()), config);
-            instance_configs.push(cfg);
-        }
+    let mut instance_configs = vec![];
+    for i in 0..num_specs {
+        let service_spec = env.get_object_array_element(specs_array, i)?;
+        let (spec, config) = parse_instance_spec(&env, service_spec, &artifact_id)?;
+        let cfg = InstanceConfig::new(spec, Some(artifact_filename.to_bytes()), config);
+        instance_configs.push(cfg);
     }
+
     Ok(instance_configs)
 }
 
-// Converts Java array of `TimeServiceSpec` to vector of `InstanceCollection`.
-fn time_service_specs_from_java_array(
+// Parses the `ServiceSpec` instance.
+//
+// `ServiceSpec` representation:
+//      String serviceName;
+//      int serviceId;
+//      byte[] configuration;
+fn parse_instance_spec(
     env: &JNIEnv,
-    executor: Executor,
-    service_specs: jobjectArray,
-) -> JniResult<Vec<InstanceCollection>> {
-    let mut instance_collection = vec![];
-    let num_configs = env.get_array_length(service_specs)?;
-    for i in 0..num_configs {
-        let service_spec_obj = env.get_object_array_element(service_specs, i)?;
+    service_spec_obj: JObject,
+    artifact_id: impl AsRef<str>,
+) -> JniResult<(InstanceSpec, Vec<u8>)> {
+    let (service_id, service_name) = get_service_id_and_name(env, service_spec_obj)?;
+    let config_params: jbyteArray = env
+        .get_field(service_spec_obj, "configuration", "[B")?
+        .l()?
+        .into_inner();
+    let config = env.convert_byte_array(config_params)?;
+    let spec = InstanceSpec::new(service_id, service_name, artifact_id).map_err(|err| {
+        JniError::from(format!(
+            "Unable to create instance specification for the service with id {}: {}",
```

Comment:
I'm getting `Unable to create instance specification for the service with id 46: invalid digit found in string` here.
I looked into that and it seems you need to attach runtime id in front of `artifact_id` so it is parsed correctly.

Reply:
Or we could do that on Java side if you want